include/vcml/models/dma/pl330.h

```diff
@@ -0,0 +1,279 @@
+
+        std::optional<T> pop(int tag) {
+            if (m_queues[tag].empty()) {
+                return std::nullopt;
+            }
+            T item = m_queues[tag].front();
+            m_queues[tag].pop_front();
+            m_tags.erase(std::find(m_tags.begin(), m_tags.end(), tag));
+            m_current_sum -= 1;
+            return std::make_optional(item);
+        }
+
+        void clear(int tag) {
+            m_queues[tag].clear();
+            m_tags.erase(std::remove(m_tags.begin(), m_tags.end(), tag),
+                         m_tags.end());
+        }
+        void remove_tagged(int tag) { clear(tag); }
+
+        void clear() {
+            m_queues.clear(); // assumes QUEUE_ITEM does not contain owning
+                              // pointers
+            m_tags.clear();
+        }
+        void reset() { clear(); }
+
+        bool empty() const { return m_tags.empty(); }
+
+        bool empty(int tag) const { return m_queues.at(tag).empty(); }
+
+        size_t size() const { return m_tags.size(); }
+
+        size_t num_free() const { return m_max_sum - m_current_sum; }
+
+        size_t size_tag(int tag) const { return m_queues[tag].size(); }
+
+    private:
+        std::unordered_map<int, std::deque<T>> m_queues;
+        std::deque<int> m_tags;
+        int m_max_sum;
+        int m_current_sum;
+    };
+
+public:
+    enum pl330_configs : u32 {
+        INSN_MAXSIZE = 6,
+        WD_TIMEOUT = 1024,
+    };
+    enum mfifo_width : u32 {
+        MFIFO_32BIT = 0b010,
+        MFIFO_64BIT = 0b011,
+        MFIFO_128BIT = 0b100,
+    };
+    enum amba_ids : u32 {
+        AMBA_PID = 0x00241330, // Peripheral ID
+        AMBA_CID = 0xb105f00d, // PrimeCell ID
+    };
+    struct queue_entry {
+        u32 data_addr;
+        u32 data_len;
+        u32 burst_len_counter;
+        bool inc;
+        bool zero_flag;
+        u32 tag;
+    };
+
+    struct mfifo_entry {
+        u8 buf;
+        u8 tag;
+    };
+
+    class channel : public module
+    {
+    public:
+        enum state : u8 {
+            STOPPED = 0x0,
+            EXECUTING = 0x1,
+            CACHE_MISS = 0x2,
+            UPDATING_PC = 0x3,
+            WAITING_FOR_EVENT = 0x4,
+            AT_BARRIER = 0x5,
+            WAITING_FOR_PERIPHERAL = 0x7,
+            KILLING = 0x8,
+            COMPLETING = 0x9,
+            FAULTING_COMPLETING = 0xE,
+            FAULTING = 0xF,
+        };
+
+        enum fault : u8 {
+            UNDEF_INSTR = 0x0,
+            OPERAND_INVALID = 0x1,
+            CH_EVNT_ERR = 0x5,
+            CH_PERIPH_ERR = 0x6,
+            CH_RDWR_ERR = 0x7,
+            MFIFO_ERR = 0xC,
+            ST_DATA_UNAVAILABLE = 0xD,
+            INSTR_FETCH_ERR = 0x10,
+            DATA_WRITE_ERR = 0x11,
+            DATA_READ_ERR = 0x12,
+            DBG_INSTR = 0x1E,
+            LOCKUP_ERR = 0x1F,
+        };
+
+        reg<u32> ftr; // channel fault type register
```

Comment:
in most IDEs this allows for hints on hover wherever the register is used. I personally find this much more ergonomic than looking up abbreviations in manuals whenever I forget them, because it allows for both short names but also quick name expansions on hover.